apollo-federation/src/merger/merger.rs

```diff
@@ -1028,8 +1029,124 @@ impl Merger {
+        let supergraph_implementations = self.merged.possible_runtime_types(dest.clone().into())?;
+
+        let mut has_key = false;
+        for (idx, source) in sources.iter() {
+            if source.is_none()
+                || self.subgraphs[*idx].is_interface_object_type(&dest.clone().into())
+            {
+                continue;
+            }
+
+            let subgraph = source.as_ref().unwrap();
+            let source_metadata = self.subgraphs[*idx].metadata();
+            let interface_pos: TypeDefinitionPosition = dest.clone().into();
+            let key_directive_name = source_metadata
+                .federation_spec_definition()
+                .key_directive_definition(&self.merged)?
+                .name
+                .clone();
+            let keys = interface_pos.get_applied_directives(subgraph.schema(), &key_directive_name);
+
+            has_key = has_key || !keys.is_empty();
+            let resolvable_key = keys.iter().find(|key| !key.arguments.is_empty());
+            if resolvable_key.is_none() {
+                continue;
+            }
+
+            let implementations_in_subgraph = subgraph
+                .schema()
+                .possible_runtime_types(dest.clone().into())?;
+            if implementations_in_subgraph.len() > supergraph_implementations.len() {
+                let missing_implementations = supergraph_implementations
+                    .iter()
+                    .filter(|implementation| !implementations_in_subgraph.contains(*implementation))
+                    .collect::<IndexSet<_>>();
+
+                self.error_reporter.add_error(CompositionError::InterfaceObjectUsageError {
+                            message: format!("Interface type \"{}\" has a resolvable key \"{}\" in subgraph \"{}\" but is missing some of the supergraph implementation types of \"{}\". Subgraph \"{}\" should define {} (and have {} implement \"{}\").",
+                                &dest.type_name,
+                                "true",
+                                &self.subgraphs[*idx].name,
+                                &dest.type_name,
+                                &self.subgraphs[*idx].name,
+                                human_readable_types(missing_implementations.iter().map(|impl_type| &impl_type.type_name)),
+                                if missing_implementations.len() > 1 { "them" } else { "it" },
+                                &dest.type_name,
+                            )
+                        });
+            }
+        }
+        Ok(has_key)
+    }
+
+    fn validate_interface_objects(
+        &mut self,
+        sources: &Sources<Subgraph<Validated>>,
+        dest: &InterfaceTypeDefinitionPosition,
+    ) -> Result<bool, FederationError> {
+        let supergraph_implementations = self.merged.possible_runtime_types(dest.clone().into())?;
+
+        for (idx, source) in sources.iter() {
+            if source.is_none()
```

Comment:
similarly as above instead of unwrapping you could
```rs
let schema = if let Some(subgraph) = source {
  subgraph.schema().schema()
} else {
  continue;
};
```